aws/resource_aws_ec2_traffic_mirror_filter_rule.go

```diff
@@ -0,0 +1,370 @@
+						"to_port": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"traffic_direction": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"ingress",
+					"egress",
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	filterId := d.Get("traffic_mirror_filter_id")
+
+	input := &ec2.CreateTrafficMirrorFilterRuleInput{
+		TrafficMirrorFilterId: aws.String(filterId.(string)),
+		DestinationCidrBlock:  aws.String(d.Get("destination_cidr_block").(string)),
+		SourceCidrBlock:       aws.String(d.Get("source_cidr_block").(string)),
+		RuleAction:            aws.String(d.Get("rule_action").(string)),
+		RuleNumber:            aws.Int64(int64(d.Get("rule_number").(int))),
+		TrafficDirection:      aws.String(d.Get("traffic_direction").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("protocol"); ok {
+		input.Protocol = aws.Int64(int64(v.(int)))
+	}
+
+	if v, ok := d.GetOk("destination_port_range"); ok {
+		input.SetDestinationPortRange(buildTrafficMirrorPortRangeRequest(v.([]interface{})))
+	}
+
+	if v, ok := d.GetOk("source_port_range"); ok {
+		input.SetSourcePortRange(buildTrafficMirrorPortRangeRequest(v.([]interface{})))
+	}
+
+	out, err := conn.CreateTrafficMirrorFilterRule(input)
+	if err != nil {
+		return fmt.Errorf("Error creating traffic mirror filter rule for %v", filterId)
+	}
+
+	d.SetId(*out.TrafficMirrorFilterRule.TrafficMirrorFilterRuleId)
+	return resourceAwsEc2TrafficMirrorFilterRuleRead(d, meta)
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	ruleId := d.Id()
+
+	var rule *ec2.TrafficMirrorFilterRule
+	filterId, filterIdSet := d.GetOk("traffic_mirror_filter_id")
+	input := &ec2.DescribeTrafficMirrorFiltersInput{}
+	if filterIdSet {
+		input.TrafficMirrorFilterIds = []*string{aws.String(filterId.(string))}
+		awsMutexKV.Lock(filterId.(string))
+		defer awsMutexKV.Unlock(filterId.(string))
+	} else {
+		awsMutexKV.Lock("DescribeTrafficMirrorFilters")
+		defer awsMutexKV.Unlock("DescribeTrafficMirrorFilters")
+	}
+
+	found := false
+	for !found {
+		out, err := conn.DescribeTrafficMirrorFilters(input)
+		if err != nil {
+			return fmt.Errorf("Error listing traffic mirror filters in the account")
+		}
+
+		if 0 == len(out.TrafficMirrorFilters) {
+			return fmt.Errorf("No traffir mirror filters found")
+		}
+
+		_, rule, err = findRuleInFilters(ruleId, out.TrafficMirrorFilters)
+		if nil == err {
+			found = true
+			break
+		} else {
+			if out.NextToken == nil {
+				break
+			} else {
+				input.NextToken = out.NextToken
+			}
+		}
+	}
+
+	if !found {
+		d.SetId("")
+		return fmt.Errorf("Rule %s not found", ruleId)
+	}
```

Comment:
The AWS Go SDK provides a paginated function ([`DescribeTrafficMirrorFiltersPages`](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeTrafficMirrorFiltersPages)) so we do not need to manually handle `NextToken` logic.

However, usually this type of logic is unnecessary. The API should be returning all children rules of the EC2 Traffic Mirror Filter in the response including the filter, not returning multiple responses of the same filter. I would suggest simplifying this logic to a single `DescribeTrafficMirrorFilters` call. We should also return `nil` when triggering resource recreation, rather than returning an error.

```go
out, err := conn.DescribeTrafficMirrorFilters(input)

if err != nil {
	return fmt.Errorf("error describing EC2 Traffic Mirror Filter (%s): %w", d.Get("traffic_mirror_filter_id").(string))
}

if out == nil || len(out.TrafficMirrorFilters) == 0 {
	log.Printf("[WARN] EC2 Traffic Mirror Filter (%s) not found, removing from state")
	d.SetId("")
	return nil
}

rule := findEc2TrafficMirrorFilterRule(ruleId, out.TrafficMirrorFilters)

if rule == nil {
	log.Printf("[WARN] EC2 Traffic Mirror Filter Rule (%s) not found, removing from state", d.Id())
	d.SetId("")
	return nil
}
```

Reply:
I believe in case of an import the read would be called without an filterId which means we would need to scan all the rules of all filters to find the proper rule. This is the reason I looped through, but ofcourse I should use the paginator.

Reply:
In that case, it might be best if the Terraform resource import ID includes both the filter ID and the filter rule ID. Otherwise the looping logic should occur in the `State` import function and leave the `Read` function simple. 👍

Reply:
Can you point me to an example using state import function?

Reply:
Example state import function, with multi-part import ID handling:

https://github.com/terraform-providers/terraform-provider-aws/blob/5cffb1d4d654d55f1d4429dfc417c1eed949e925/aws/resource_aws_api_gateway_authorizer.go#L24-L36